### ISSUE TITLE
Uniswap v4 support

### ISSUE THREAD

**Issue by ErikBjare** [uniswap-python/uniswap-python#337]
Uniswap v4 was recently announced: https://blog.uniswap.org/uniswap-v4

I'm not sure how much time I'll have to work on it myself, but I am at minimum open to helping out with reviewing PRs etc.

Not sure if we should go for a full rewrite this time and rip out the v1/v2/v3 support (to clean up the codebase), perhaps ship it as a separate package `uniswap-python-v4`.

I assume we will be able to acquire funding for this, given the Uniswap Grants Program (UGP) was satisfied with our work in the past. (as was [Adam](https://twitter.com/UniswapPython/status/1624024175781285891))

@shanefontaine @liquid-8

**Comment by liquid-8** [comment]
I've just read whitepaper, so it's kinda challenging to determine further development. From what I understand, they intend to drop old pools in favor of singleton contract. Most probably, liquidity providers will tend to migrate there as uni team declares pretty much attractive features for this contract. Hence, uni v2 and v3 pools will follow v1. I think we should wait for testnet deployment. At that moment it will be much easier to understand what way we should consider.

Regarding participation, I'd like to but I literally reside in Ukraine, so it's kinda hard to predict.

**Comment by ErikBjare** [comment]
@liquid-8 Oh shit, didn't realize. If I can help in anyway, don't hesitate to ask.

I'm also eager for a testnet deployment, just wanted to send out notice well in advance :)

**Comment by ErikBjare** [comment]
Someone YOLO-deployed v4 to Goerli: https://twitter.com/_prestwich/status/1669851049757401088

**Comment by liquid-8** [comment]
That's fine. It is possible now to make some drafts, at least. Also, I share your idea to ship that as a separate package. I doubt that we have to pay attention to the older versions from now on. However, there are plenty of v2/v3 clones; therefore, there is no need to purge previous versions' support from the codebase.

**Comment by liquid-8** [comment]
@ErikBjare Check your e-mail, pls. Sent from moy***dro@gmail.com

**Comment by ErikBjare** [comment]
Also ping @KeremP

**Comment by KeremP** [comment]
I agree a separate package is a good approach here - can leave the older versions untouched as Im sure most would try to migrate to v4 (given the optionality hooks could provide). Haven't dived into the whitepaper yet but definitely intrigued. Ill probably set aside some time and take a look this week.

**Comment by liquid-8** [comment]
> Someone YOLO-deployed v4 to Goerli: https://twitter.com/_prestwich/status/1669851049757401088

Also, test deployment from Uni team at Sepolia
https://sepolia.etherscan.io/address/0xb69e301c95d4399a5b4777a193c518f77f60b841

**Comment by snarflakes** [comment]
As an alternative to incorporating all the changes into the existing build, I recently built a Python-based script that makes it easy to execute **SWAP_EXACT_IN** swaps using Uniswap v4 Universal Router on L2. 
- 
- Provides an easy-to-use example of **how to integrate Uniswap v4 Universal Router decoding in real swaps**.
- Reduces the complexity of constructing transactions manually.
- Works currently with **v3 pools** on all **L2 networks** (including Base) (soon v4 pools once enough token pairs are available on all chains)

Here’s the repo: **[https://github.com/snarflakes/uniswap-python-swapper](https://github.com/snarflakes/uniswap-python-swapper)**

I’d love to know if you think this could be helpful for your users? Or build off of it?

Thanks for building such a great tool, i've used it extensively—looking forward to your thoughts!  

Best,  
Snar

**Comment by ErikBjare** [comment]
Support for Uniswap V4 was merged today in https://github.com/uniswap-python/uniswap-python/pull/418, will release v0.8.0 with it soon.
